arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,832 @@
+
+  // If we have not already converted this frame, try to convert it
+  if (!d->current_image_memory && d->f_frame->data[0] != 0)
+  {
+    int width = enc->width;
+    int height = enc->height;
+    int num_pixels = 3;
+    vital::image_pixel_traits pixel_trait = vital::image_pixel_traits_of<unsigned char>();
+    bool direct_copy;
+
+    // If the pixel format is not recognized by then convert the data into RGB_24
+    switch (enc->pix_fmt)
+    {
+      case AV_PIX_FMT_GRAY8:
+      {
+        num_pixels = 1;
+        direct_copy = true;
+        break;
+      }
+      case AV_PIX_FMT_RGBA:
+      {
+        num_pixels = 4;
+        direct_copy = true;
+        break;
+      }
+      case AV_PIX_FMT_MONOWHITE:
+      case AV_PIX_FMT_MONOBLACK:
+      {
+        num_pixels = 1;
+        pixel_trait = vital::image_pixel_traits_of<bool>();
+        direct_copy = true;
+        break;
+      }
+      default:
+      {
+        direct_copy = false;
+      }
+    }
+    if (direct_copy)
+    {
+      int size = avpicture_get_size(enc->pix_fmt, width, height);
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+
+      AVPicture frame;
+      avpicture_fill(&frame, (uint8_t*)d->current_image_memory->data(), enc->pix_fmt, width, height);
+      av_picture_copy(&frame, (AVPicture*)d->f_frame, enc->pix_fmt, width, height);
+    }
+    else
+    {
+      int size = width * height * num_pixels;
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+
+      d->f_software_context = sws_getCachedContext(
+        d->f_software_context,
+        width, height, enc->pix_fmt,
+        width, height, AV_PIX_FMT_RGB24,
+        SWS_BILINEAR,
+        NULL, NULL, NULL);
+
+      if (!d->f_software_context)
+      {
+        LOG_ERROR(this->logger(), "Couldn't create conversion context");
+        return nullptr;
+      }
+
+      AVPicture rgb_frame;
+      avpicture_fill(&rgb_frame, (uint8_t*)d->current_image_memory->data(), AV_PIX_FMT_RGB24, width, height);
+
+      sws_scale(d->f_software_context,
+        d->f_frame->data, d->f_frame->linesize,
+        0, height,
+        rgb_frame.data, rgb_frame.linesize);
+    }
+
+    vital::image image(
+      d->current_image_memory,
+      d->current_image_memory->data(),
+      width, height, num_pixels,
+      num_pixels, num_pixels * width, num_pixels * width*height
+    );
+    d->current_image = std::make_shared<vital::simple_image_container>(vital::simple_image_container(image));
+  }
+
+  return d->current_image;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::timestamp
+ffmpeg_video_input
+::frame_timestamp() const
+{
+  if (!this->good())
+  {
+    return {};
+  }
+
+  // We don't always have all components of a timestamp, so start with
+  // an invalid TS and add the data we have.
+  kwiver::vital::timestamp ts;
+  ts.set_frame(d->frame_number() + d->f_frame_number_offset + 1);
```

Comment:
Do not use this `frame_number()` function to compute the current frame.   It is not reliable.  We should keep an internal counter indicating the last frame we stepped to and return that instead.

Reply:
@mleotta: Do you want to solve this before landing the branch ?

Reply:
Let's land the branch as is and the follow up with improvements